src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -154,41 +169,119 @@ def bsp_resolve_field_factory(
         )
-    resolve = list(resolves)[0]
-    scala_version = scala.version_for_resolve(resolve)
+    (resolve,) = resolves
 
+    scala_version = scala.version_for_resolve(resolve)
     scala_runtime = await Get(
         MaterializeScalaRuntimeJarsResult, MaterializeScalaRuntimeJarsRequest(scala_version)
     )
 
+    #
+    # Extract the JDK paths from an lawful-evil process so we can supply it to the IDE
+    #
+
+    # The maximum JDK version will be compatible with all the specified targets
+    jdk_requests = [JdkRequest.from_field(version) for version in jdk_versions]
+    jdk_request = max(jdk_requests, key=_jdk_request_sort_key(jvm))
+
+    jdk, readlink, = await MultiGet(
+        Get(JdkEnvironment, JdkRequest, jdk_request),
+        Get(ReadlinkBinary, ReadlinkBinaryRequest()),
+    )
+
+    cmd = "leak_paths.sh"
+    leak_jdk_sandbox_paths = textwrap.dedent(
+        f"""\
+        # Script to leak JDK cache paths out of Coursier sandbox so that BSP can use them.
+
+        {readlink.path} {jdk.coursier.cache_dir}
+        {jdk.java_home_command}
+        """
+    )
+    leak_sandbox_path_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    cmd,
+                    leak_jdk_sandbox_paths.encode("utf-8"),
+                    is_executable=True,
+                ),
+            ]
+        ),
+    )
+
+    leaked_paths = await Get(
+        ProcessResult,
+        Process(
+            [
+                bash.path,
+                cmd,
+            ],
+            input_digest=leak_sandbox_path_digest,
+            immutable_input_digests=jdk.immutable_input_digests,
+            env=jdk.env,
+            use_nailgun=(),
+            description="Leak JDK cache paths for BSP",
+            append_only_caches=jdk.append_only_caches,
+        ),
+    )
+
+    cache_dir, jdk_home = leaked_paths.stdout.decode().strip().split("\n")
+
+    _, _, suffix = jdk_home.partition(jdk.coursier.cache_dir)
+    coursier_java_home = cache_dir + suffix
+
     scala_jar_uris = tuple(
         build_root.pathlib_path.joinpath(".pants.d/bsp").joinpath(p).as_uri()
         for p in scala_runtime.content.files
     )
 
+    jvm_build_target = JvmBuildTarget(
+        Path(coursier_java_home).as_uri(), f"1.{jdk.jre_major_version}"
+    )
+
     return BSPBuildTargetsMetadataResult(
         metadata=ScalaBuildTarget(
             scala_organization="org.scala-lang",
             scala_version=scala_version,
             scala_binary_version=".".join(scala_version.split(".")[0:2]),
             platform=ScalaPlatform.JVM,
             jars=scala_jar_uris,
+            jvm_build_target=jvm_build_target,
         ),
         digest=scala_runtime.content.digest,
     )
 
 
+def _jdk_request_sort_key(
+    jvm: JvmSubsystem,
+) -> Callable[[JdkRequest,], tuple[int, ...]]:
+    def sort_key_function(request: JdkRequest) -> tuple[int, ...]:
+        if request == JdkRequest.SYSTEM:
```

Comment:
What is there are multiple JDKs and the system JDK is _not_ selected? Maybe move this check to the call site actually sorting the results?

Reply:
`JdkEnvironment` should be filled out (including `jre_major_version`) for system JDKs? Am I missing something?

Reply:
Oh, that's probably right 🤦